### PR TITLE
Refactor the backbone builders of detection

### LINKED ISSUE #4676: Refactor Models builders to make them reusable on Prototype
### 🚀 The feature

The current model builders are hard to reuse from the prototype area, leading to large amounts of copy-pasting. It would be good to refactor the space in a BC manner so that we can reuse as many parts as we can:
- [x] Segmentation models:
  - Update them to follow the standard convention of defining the builders on the same file as their class.
  - Make private model builders reusable from prototype.
  - Clean up the implementations and expose publicly variables that are currently pushed to kwargs.
- [x] Detection models:
  - Make private model builders reusable from prototype.
  - Refactor the VGG16 Features weights mechanism on SSD to work well with the proposed weights prototype. 

### Motivation, pitch

-

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by datumbox**
Fixes #4676

Implements some of the same ideas on the detection space. More specifically, we move the CNN backbone initialization on the model builder and pass it directly to the method that implements the feature extractor for the specific task.

This PR will enable the work on #4611. Changes are BC, though we are forced to update an expected file because of the different order of initialization. See https://github.com/pytorch/vision/pull/4656#issuecomment-946955438 for details.

Accuracy remains the same for all pre-trained models.

### CODE REVIEW

**Comment by datumbox** [comment]
@NicolasHug Thanks for the help!

> Do you think this is something worth mentioning in the future release notes, or nah?

Hm... I don't think we have done this before. This is probably because PyTorch doesn't guarantee getting the same random behaviour across versions. @fmassa any thoughts on this based on history?

We have 3 options here:
1. Say nothing.
2. Just mention it on the notes.
3. Ensure that the behaviour is identical with before. For that I just need to initialize the `LastLevelP6P7(256, 256)` before the backbone and we are good to go.

I don't mind going for any of the above solutions. I'll merge now but happy to follow up with option 3 if needed.

**Comment by datumbox** [comment]
I've tested again all models and their metrics look fine. I'm merging this.
